Code/Editor/EditorFramework/CodeGen/CompilerPreferencesWidget.cpp

```diff
@@ -0,0 +1,167 @@
+
+  auto gridLayout = new QGridLayout();
+  gridLayout->setColumnStretch(0, 1);
+  gridLayout->setColumnStretch(1, 0);
+  gridLayout->setColumnMinimumWidth(1, 5);
+  gridLayout->setColumnStretch(2, 2);
+  gridLayout->setContentsMargins(0, 0, 0, 0);
+  gridLayout->setSpacing(0);
+
+  ezStringBuilder fmt;
+  QLabel* versionText = new QLabel(ezMakeQString(ezFmt("This SDK was compiled with {} version {}. Select a compatible compiler.",
+    ezCppProject::CompilerToString(ezCppProject::GetSdkCompiler()),
+    ezCppProject::GetSdkCompilerMajorVersion())
+                                                   .GetText(fmt)));
+  versionText->setWordWrap(true);
+  gridLayout->addWidget(versionText, 0, 0, 1, 3);
+
+  gridLayout->addWidget(new QLabel("Compiler Preset"), 1, 0);
+  gridLayout->addWidget(m_pCompilerPreset, 1, 2);
+  m_pGroupLayout->addLayout(gridLayout);
+}
+
+ezQtCompilerPreferencesWidget::~ezQtCompilerPreferencesWidget() = default;
+
+void ezQtCompilerPreferencesWidget::SetSelection(const ezHybridArray<ezPropertySelection, 8>& items)
+{
+  ezQtScopedUpdatesDisabled _(this);
+
+  ezQtPropertyTypeWidget::SetSelection(items);
+
+  if (m_pTypeWidget)
+  {
+    const auto& selection = m_pTypeWidget->GetSelection();
+
+    EZ_ASSERT_DEBUG(selection.GetCount() == 1, "Expected exactly one object");
+    auto pObj = selection[0].m_pObject;
+
+    ezEnum<ezCompiler> eCompiler;
+    bool bIsCustomCompiler;
+    ezString sCCompiler, sCppCompiler;
+
+    {
+      ezVariant varCompiler, varIsCustomCompiler, varCCompiler, varCppCompiler;
+
+      m_pObjectAccessor->GetValue(pObj, "Compiler", varCompiler).AssertSuccess();
+      m_pObjectAccessor->GetValue(pObj, "CustomCompiler", varIsCustomCompiler).AssertSuccess();
+      m_pObjectAccessor->GetValue(pObj, "CCompiler", varCCompiler).AssertSuccess();
+      m_pObjectAccessor->GetValue(pObj, "CppCompiler", varCppCompiler).AssertSuccess();
+
+      eCompiler.SetValue(static_cast<ezCompiler::StorageType>(varCompiler.Get<ezInt64>()));
+      bIsCustomCompiler = varIsCustomCompiler.Get<decltype(bIsCustomCompiler)>();
+      sCCompiler = varCCompiler.Get<decltype(sCCompiler)>();
+      sCppCompiler = varCppCompiler.Get<decltype(sCppCompiler)>();
+    }
+
+    ezInt32 selectedIndex = -1;
+    const auto& machineSpecificCompilers = ezCppProject::GetMachineSpecificCompilers();
+    // first look for non custom compilers
+    for (ezUInt32 i = 0; i < machineSpecificCompilers.GetCount(); ++i)
+    {
+      const auto& curCompiler = machineSpecificCompilers[i];
+      if (curCompiler.bIsCustom == false && curCompiler.eCompiler == eCompiler && curCompiler.sCCompiler == sCCompiler && curCompiler.sCppCompiler == sCppCompiler)
+      {
+        selectedIndex = static_cast<int>(i);
+        break;
+      }
+    }
+
+    if (selectedIndex == -1)
+    {
+      // If we didn't find a system default compiler, look for custom compilers next
```

Comment:
A custom compiler is a compiler where the user manually specified the compiler path. This code is not about what we prefer though. For simplicity there is a dropdown in the widget which doesn't get serialized (because it's contents can change on very editor restart, so it doesn't make sense to serialize it). What this code does is, it reconstructs the dropdowns selected index from the information we did actually serialize. So this has nothing to do with what we prefer in terms of a compiler.

Reply:
We never throw away the users settings. Even if the compiler he configured earlier is no longer avaiable, we still retain the settings he made.